src/main/java/com/minecolonies/coremod/util/StructureWrapper.java

```diff
@@ -162,6 +194,46 @@ public void rotate(final int times, @NotNull final World world, @NotNull final B
      *
      * @param pos coordinates
      */
+    private boolean checkForFreeSpace(@NotNull final BlockPos pos)
+    {
+        setLocalPosition(pos);
+        //structure.getBlockInfo()[0].pos
+        for (int j = 0; j < structure.getHeight(); j++)
+        {
+            for (int k = 0; k < structure.getLength(); k++)
+            {
+                for (int i = 0; i < structure.getWidth(); i++)
+                {
+                    @NotNull final BlockPos localPos = new BlockPos(i, j, k);
+
+                    final BlockPos worldPos = pos.add(localPos);
+
+                    if(worldPos.getY() <= pos.getY() && !world.getBlockState(worldPos.down()).getMaterial().isSolid())
+                    {
+                        return false;
+                    }
+
+                    final IBlockState worldState = world.getBlockState(worldPos);
+                    if(worldState.getBlock() == Blocks.BEDROCK)
+                    {
+                        return false;
+                    }
+
+                    if(worldPos.getY() > pos.getY() && worldState.getBlock() != Blocks.AIR)
+                    {
+                        return false;
+                    }
+                }
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Place a structure into the world.
+     *
+     * @param pos coordinates
+     */
     private void placeStructure(@NotNull final BlockPos pos)
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)